### ISSUE TITLE
Css/js resources should be cached by browsers

### ISSUE THREAD

**Issue by kelson42** [openzim/zimfarm#537]
Proper HTTP response headers should be set

**Comment by rgaudin** [comment]
More details please ; those resources are served as static files with proper `Modified` and `Etag` that any modern browser caches properly (devtools shows it in Network tab).

**Comment by kelson42** [comment]
I believe that without `Cache-Control` header, the browser makes always a request to the server. The `Etag` or `last-modified` are only there to avoid a full `GET`, but it should still make a `HEAD` request each time. Probable other impact: no proxy will ever cache anything.

**Comment by rgaudin** [comment]
Etag were specifically designed for proxy.

I understand the other point though. Let me think about appropriate values for those as users (you) always want changes immediately visible ;)

**Comment by mahakporwal02** [comment]
@rgaudin @kelson42 should I work on this issue?

**Comment by rgaudin** [comment]
We need to defined a value for this setting: how long should the JS and CSS be cached. Given the whole site is a JS app, it's sortof important… but it's also an *internal* app used by a small group of persons who knows to force-refresh on issues.
@kelson42 ?

**Comment by kelson42** [comment]
@rgaudin The cache duration shoukd be long, maybe a month. Bit would be great to add etag based on web server start time and some like “?etag” to the css/js dependences to avoid a force refresh on client side.

**Comment by rgaudin** [comment]
@kelson42 that sounds non trivial given we are in a server-less environment: we build our source code into a large JS bundle that is later served by nginx.
nginx [uses the timestamp of the file on the FS to calculate its Etag](http://lxr.nginx.org/source/xref/nginx/src/http/ngx_http_core_module.c?r=7638%3A681b78a98a52#1656) and that sounds difficult for our use case: would imply modifying the HTML part post-build or even post-COPY step.

Anyway, that would not be necessary given our toolchain already creates and links to unique filenames/url based on a hash of the content (ex `<link href="/css/app.3938340f.css"`).

So to me, we should just send a long-lasting `Cache-Control` header. Your suggestion of a month seems good).

**Comment by kelson42** [comment]
@mahakporwal02 I guess you have now everything you need to make a fix.

**Comment by benoit74** [comment]
To me, there is a bad understanding of browser behavior in this issue.

Browsers never send `HEAD` requests. They do send `GET` requests with `If-None-Match: "etagvalue"` and `If-Modified-Since: xxxx`, when `Cache-Control: no-cache` was returned on last query, or when `Cache-Control` was not returned and their internal heuristic decides it is worth to revalidate content freshness. The server will then respond with `304: Not Modified` or with `200` and the updated content.

No `Cache-Control` header is hence not a major concern for JS and CSS in our use case, especially since JS and CSS names are updated everytime we rebuild the app. It simply means that browsers will use their own heuristics while we could have cached a bit more aggressively. In other words, we will have sporadic `GET` requests which end up with a `304`. Not a big deal.

On the other end of the spectrum, we do have an issue with HTML content is also cached by the browser, while it should not. 
Everytime we redeploy, a new HTML is supposed to be returned with new hashed JS / CSS filename ... but since it is cached old HTML is still used, using old JS / CSS ... which are not available. Same for static JSON files like `config.json`.

This is not a big concern because browser default heuristics are kinda conservative (e.g. Firefox always re-validate cached entries at least once per day). But it is not totally correct.

@elfkuzco can you confirm my understanding by testing locally with two subsequent "normal" frontend Docker images and normal cache behavior in your browser (i.e. not hitting force refresh)?

I suspect we should configure Nginx properly to set `Cache-Control: no-cache` on HTML and JSON content-type, `Cache-Control: public, max-age=2592000, immutable` on JS and CSS, and `Cache-Control: public, max-age=2592000` on images as well.

**Comment by elfkuzco** [comment]
Yeah, indeed the JS and CSS files aren't cached and are always fetched.

**Comment by benoit74** [comment]
This is not what I said. On my end, they are cached, but not based on a specific condition we have set

**Comment by elfkuzco** [comment]
Switched to chrome and yeah it was cached. The browser didn't send any `If-None-Match` or `If-Modified-Since`. The json config file is however cached along with the HTML (as you pointed out). Asides that, every other thing is similar.

**Comment by benoit74** [comment]
OK, then please move on with following modification in nginx:
- `Cache-Control: no-cache` on HTML and JSON content-types
- `Cache-Control: public, max-age=2592000, immutable` on JS and CSS content types
- `Cache-Control: public, max-age=2592000` on images content-types
